src/shared/utilities/timeoutUtils.ts

```diff
@@ -200,28 +238,30 @@ export function waitTimeout<T, R = void, B extends boolean = true>(
         promise = promise()
     }
 
-    return Promise.race([promise, timeout.timer])
-        .then(obj => {
-            if (obj !== undefined) {
-                return obj
-            }
-            if ((opt.allowUndefined ?? true) !== true) {
-                throw new Error(TIMEOUT_UNEXPECTED_RESOLVE)
-            }
-            return undefined as any
-        })
-        .catch(err => {
-            if (opt.onExpire && (err as Error).message === TIMEOUT_EXPIRED_MESSAGE) {
-                return opt.onExpire()
-            }
-            if (opt.onCancel && (err as Error).message === TIMEOUT_CANCELLED_MESSAGE) {
-                return opt.onCancel()
-            }
-            throw err
-        })
+    const result = await Promise.race([promise, timeout.promisify()])
+        .catch(e => (e instanceof Error ? e : new Error(`unknown error: ${e}`)))
         .finally(() => {
             if ((opt.completeTimeout ?? true) === true) {
                 timeout.complete()
             }
         })
+
+    if (result instanceof Error) {
+        if (opt.onExpire && TimeoutError.isExpired(result)) {
+            return opt.onExpire()
+        }
+        if (opt.onCancel && TimeoutError.isCancelled(result)) {
+            return opt.onCancel()
+        }
+        throw result
+    }
+
+    if (result !== undefined) {
+        return result
+    }
```

Comment:
Can you combine this with the final statement? Just `return result` or something.